File: src/test/java/mobi/hsz/idea/gitignore/util/MatcherUtilTest.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 hsz Jakub Chrzanowski <jakub@hsz.mobi>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package mobi.hsz.idea.gitignore.util;

import mobi.hsz.idea.gitignore.Common;
import org.junit.Test;

import java.util.regex.Pattern;

public class MatcherUtilTest extends Common<MatcherUtil> {

    @Test
    public void testMatch() {
        final Pattern pattern = Pattern.compile("foo");
        final MatcherUtil util = new MatcherUtil();

        assertFalse(util.match(null, null));
        assertFalse(util.match(null, "foo"));
        assertFalse(util.match(pattern, null));
        assertFalse(util.match(pattern, "fo"));
        assertTrue(util.match(pattern, "foo"));
        assertTrue(util.match(pattern, "xfooy"));
    }

    @Test
    public void testMatchAllParts() {
        final String[] partsA = new String[]{"foo"};
        final String[] partsB = new String[]{"foo", "bar"};

        assertFalse(MatcherUtil.matchAllParts(null, null));
        assertFalse(MatcherUtil.matchAllParts(null, "foo"));
        assertFalse(MatcherUtil.matchAllParts(partsA, null));
        assertFalse(MatcherUtil.matchAllParts(partsA, "fo"));
        assertTrue(MatcherUtil.matchAllParts(partsA, "foo"));
        assertTrue(MatcherUtil.matchAllParts(partsA, "xfooy"));
        assertFalse(MatcherUtil.matchAllParts(partsB, "xfooxba"));
        assertTrue(MatcherUtil.matchAllParts(partsB, "xfooxbar"));
    }

    @Test
    public void testMatchAnyPart() {
        final String[] partsA = new String[]{"foo"};
        final String[] partsB = new String[]{"foo", "bar"};

        assertFalse(MatcherUtil.matchAnyPart(null, null));
        assertFalse(MatcherUtil.matchAnyPart(null, "foo"));
        assertFalse(MatcherUtil.matchAnyPart(partsA, null));
        assertFalse(MatcherUtil.matchAnyPart(partsA, "fo"));
        assertTrue(MatcherUtil.matchAnyPart(partsA, "foo"));
        assertTrue(MatcherUtil.matchAnyPart(partsA, "xfooy"));
        assertTrue(MatcherUtil.matchAnyPart(partsB, "xfooxba"));
        assertTrue(MatcherUtil.matchAnyPart(partsB, "xfooxbar"));
    }

    @Test
    public void testGetParts() {
        Pattern pattern = Pattern.compile("foo[ba]rbuz.*hi");

        assertEquals(MatcherUtil.getParts(null).length, 0);

        assertEquals(MatcherUtil.getParts(pattern).length, 2);

        pattern = Pattern.compile("$$_!@[fd]");
        assertEquals(MatcherUtil.getParts(pattern).length, 0);
    }
}

File: src/test/kotlin/mobi/hsz/idea/gitignore/util/MatcherUtilTest.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package mobi.hsz.idea.gitignore.util

import com.intellij.openapi.components.service
import junit.framework.TestCase
import mobi.hsz.idea.gitignore.Common
import mobi.hsz.idea.gitignore.services.IgnoreMatcher
import org.junit.Test
import java.util.regex.Pattern

class MatcherUtilTest : Common<MatcherUtil?>() {

    @Test
    fun testMatch() {
        val pattern = Pattern.compile("foo")
        val matcher = project.service<IgnoreMatcher>()

        TestCase.assertFalse(matcher.match(null, null))
        TestCase.assertFalse(matcher.match(null, "foo"))
        TestCase.assertFalse(matcher.match(pattern, null))
        TestCase.assertFalse(matcher.match(pattern, "fo"))
        TestCase.assertTrue(matcher.match(pattern, "foo"))
        TestCase.assertTrue(matcher.match(pattern, "xfooy"))
    }

    @Test
    fun testMatchAllParts() {
        val partsA = arrayOf<String?>("foo")
        val partsB = arrayOf<String?>("foo", "bar")
        TestCase.assertFalse(MatcherUtil.matchAllParts(null, null))
        TestCase.assertFalse(MatcherUtil.matchAllParts(null, "foo"))
        TestCase.assertFalse(MatcherUtil.matchAllParts(partsA, null))
        TestCase.assertFalse(MatcherUtil.matchAllParts(partsA, "fo"))
        TestCase.assertTrue(MatcherUtil.matchAllParts(partsA, "foo"))
        TestCase.assertTrue(MatcherUtil.matchAllParts(partsA, "xfooy"))
        TestCase.assertFalse(MatcherUtil.matchAllParts(partsB, "xfooxba"))
        TestCase.assertTrue(MatcherUtil.matchAllParts(partsB, "xfooxbar"))
    }

    @Test
    fun testMatchAnyPart() {
        val partsA = arrayOf<String?>("foo")
        val partsB = arrayOf<String?>("foo", "bar")
        TestCase.assertFalse(MatcherUtil.matchAnyPart(null, null))
        TestCase.assertFalse(MatcherUtil.matchAnyPart(null, "foo"))
        TestCase.assertFalse(MatcherUtil.matchAnyPart(partsA, null))
        TestCase.assertFalse(MatcherUtil.matchAnyPart(partsA, "fo"))
        TestCase.assertTrue(MatcherUtil.matchAnyPart(partsA, "foo"))
        TestCase.assertTrue(MatcherUtil.matchAnyPart(partsA, "xfooy"))
        TestCase.assertTrue(MatcherUtil.matchAnyPart(partsB, "xfooxba"))
        TestCase.assertTrue(MatcherUtil.matchAnyPart(partsB, "xfooxbar"))
    }

    @Test
    fun testGetParts() {
        var pattern = Pattern.compile("foo[ba]rbuz.*hi")
        TestCase.assertEquals(MatcherUtil.getParts(null).size, 0)
        TestCase.assertEquals(MatcherUtil.getParts(pattern).size, 2)
        pattern = Pattern.compile("$$!@[fd]")
        TestCase.assertEquals(MatcherUtil.getParts(pattern).size, 0)
    }
}
